datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaTransportProvider.java

```diff
@@ -136,6 +140,40 @@ private int getSourcePartitionFromEvent(BrooklinEnvelope event) {
         event.getMetadata().getOrDefault(BrooklinEnvelopeMetadataConstants.SOURCE_PARTITION, "-1"));
   }
 
+  @Override
+  public void broadcast(String destinationUri, DatastreamProducerRecord record,
+      SendCallback onBroadcastComplete, SendCallback onEventComplete) {
+    Validate.isTrue(record.isBroadcastRecord(), "Trying to broadcast a non-broadcast type record.");
+
+    // Currently destination topic partition count will be queried from datastream destination metadata. This implies
+    // a restriction that datastream destination partition should be updated for broadcast to work correctly.
+    int partitionCount = _datastreamTask.getDatastreamDestination().getPartitions();
+    String topicName = KafkaTransportProviderUtils.getTopicName(destinationUri);
+
+    LOG.debug("Broadcasting record {} to all {} partitions of destination {}", record, partitionCount, destinationUri);
+    int partition = 0;
+    List<Integer> sentToPartitions = new ArrayList<>();
+    try {
+      for (; partition < partitionCount; partition++) {
+        record.setPartition(partition);
+        send(destinationUri, record, ((metadata, exception) -> {
+          if (exception != null) {
+            LOG.error("Failed to broadcast record {} to partition {}", record, metadata.getPartition());
+          } else {
+            LOG.debug("Sent broadcast record {} to partition {}", record, metadata.getPartition());
+            onEventComplete.onCompletion(metadata, exception);
+          }
+        }));
+        sentToPartitions.add(partition);
+      }
+      doOnBroadcastCallback(record, onBroadcastComplete, sentToPartitions, topicName, partitionCount, null);
+    } catch (DatastreamRuntimeException ex) {
+      LOG.error("Broadcast send failed for record {} at partition {}/{} because of exception: {} ",
+          record, partition, partitionCount, ex);
+      doOnBroadcastCallback(record, onBroadcastComplete, sentToPartitions, topicName, partitionCount, ex);
```

Comment:
Looks like the only exceptions we pass to the onBroadcastCallback is if send() throws (e.g. serde exceptions). But that doesn't capture the async exceptions that send() provides via callbacks (e.g. delivery timeouts). So we're swallowing any async exceptions.

Reply:
The onEventCallback is where it would be handled.